documentation/blog.md

```diff
@@ -0,0 +1,27 @@
+
+Config files API (CFA) is a library written in ruby intended for accessing various configuration files. It is structured into several layers and creates an internal abstraction of configuration file. It has again been designed and developed in SUSE's YaST team. However this time it uses (or can use) third party parsers. CFA provides location transparency for the parser on the bottom layer and unific API for application on the top one.
+
+Lets go through layers in details.
+
+## First layer ##
+
+Is responsible for accessing configuration files itself. In simplest case it accesses local configuration files, but it can be adapted to access remote, chrooted or memory files too.
+
+## Second layer ##
+
+This layer parses the configuration file which was loaded by the underlying layer. It knows structure of the file and transforms it into abstract representation. The library typically uses external tool for parsing here (e.g. Augeas). So, if the external tools has a specific requirements, it has to be satisfied to get things working. For example if the Augeas is in use you need to provide him with a proper lens to parse the particular configuration file.
+
+## Third layer ##
+
+Last layer creates a model of the configuration file - an API for accessing configuration from an application. It basically creates "an abstraction on top of another abstraction". It means that it unifies use of various tools which can be used for accessing and parsing various configuration files.
+
+## Limitations ##
+
+The approach as described above has several limitations.
+
+Firstly you need to feed a parser too (e.g. provide lenses for augeas as described above). So, you at least has to know which parser is used for parsing which file and provide him with all stuff needed for parsing of file of your interest. This is especially important in case of nonstandard / custom configuration files.
+
+Second limitation is that the library and / or parsers on second layer use an abstraction for representing configuration file. This abstraction transform a configuration file into a model and establishes a relation between the file and the model. This relation is not bijective (typicall example is Augeas). It means that some irrelevant pieces of the configuration file are not represented in the model. For example some spaces can be left out if these are not needed from syntactic point of view. This can consequently lead to loose of custom padding.
```

Comment:
in general I do not know currently parser that store whitespaces. but good to mention it, that it is often limitation. Less smart parsers even ignore comments.